### ISSUE TITLE
Infection fails with "Allowed memory size ... exhausted" for a project with a really big coverage

### ISSUE THREAD

**Issue by maks-rafalko** [infection/infection#705]
| Question    | Answer
| ------------| ---------------
| Infection version | 0.13.2
| Test Framework version | PHPUnit
| PHP version | 7.2
| Github Repo | https://github.com/vimeo/psalm 

Extracted from https://twitter.com/sorsoup/status/1140021198765658118

> Infection seems to be crashing when I try to run it to check the test suite of Psalm - output at (link: https://circleci.com/gh/bdsl/psalm/232) circleci.com/gh/bdsl/psalm/…
> 
> Don't know if I'm doing anything wrong.

Command line:

```bash
php -d memory_limit=12G ~/sites/infection/bin/infection --coverage=build/phpunit --only-covered --threads=4
```

Output:

```bash
     ____      ____          __  _
    /  _/___  / __/__  _____/ /_(_)___  ____
    / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
  _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
 /___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Running initial test suite...

PHPUnit version: 7.5.12

    1 [============================] < 1 sec

Generate mutants...

Processing source code files:   0/543

PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /infection/src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php on line 170
```

So it fails with when the coverage data is being accumulated in the following lines of the code (depending on how much `memory_limit` is set

* on [`src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php:170`](https://github.com/infection/infection/blob/b9ebc49f2b5fd923b848a2ea035f90577d02bc5c/src/TestFramework/PhpUnit/Coverage/CoverageXmlParser.php#L170)
* on [`src/TestFramework/Coverage/CodeCoverageData.php:220`](https://github.com/infection/infection/blob/b9ebc49f2b5fd923b848a2ea035f90577d02bc5c/src/TestFramework/Coverage/CodeCoverageData.php#L220)

Unfortunately, code coverage takes 2+ hours for Psalm, so here is a compressed archive of needed coverage files: https://www.dropbox.com/s/66tvd0tctfbdki3/phpunit.bz?dl=0 for commit https://github.com/vimeo/psalm/commit/4c57c67e2f1b821e2ef8aa0b00448ff24ce92201

**Comment by bdsl** [comment]
> Unfortunately, code coverage takes 2+ hours for Psalm,

I guess this is using xdebug. In the Psalm CircleCI pipeline coverage data is collected using [PCOV](https://github.com/krakjoe/pcov), and [takes around ten minutes](https://circleci.com/gh/vimeo/psalm/58).

**Comment by bdsl** [comment]
Do you know why I don't see the message about allowed memory size? I did see that before I added `-d memory_limit=4G` but it isn't there in my recent builds, e.g. https://circleci.com/gh/bdsl/psalm/353

**Comment by sanmai** [comment]
> Allowed memory size of 268435456 bytes exhausted

That's 256 Mb for short. It seems like `memory_limit` didn't stick through.

**Comment by sanmai** [comment]
@bdsl if you ssh into that image, will be there more details? It may be hitting some outside limits, like if that VM has no swap, and not 4 Gb of RAM.

**Comment by maks-rafalko** [comment]
> That's 256 Mb for short. It seems like `memory_limit` didn't stick through.

I've copied it from local machine. When I run it with `php -dmemory_limit=12G` it fails with `Allowed memory size of 12884901888 bytes exhausted`. So there is no issue with passing `memory_limit` to the infection process. At least locally.

The issue is indeed in the very big coverage data files (~2Gb in total).

**Comment by maks-rafalko** [comment]
I've [created a PR](https://github.com/infection/infection/pull/708) with some minor changes that allowed me to run Infection for Psalm with `memory_limit=12G` without that fatal error.

This is not a final fix, just a few minor (but still useful) improvements.

@bdsl I'm pretty sure, you need to run Infection for Psalm [only for changed files](https://infection.github.io/guide/how-to.html#How-to-run-Infection-only-for-changed-files). This will *dramatically* reduce the mutation time. I'm not sure Infection is ready for such many mutations:

```bash
Processing source code files: 543/543
Creating mutated files and processes:    0/20516
```

You can see how we do it [for Infection itself.](https://github.com/infection/infection/blob/e34f2b59d417da09f56a39a9c9119c41b4d913c3/.ci/travis-functions.sh#L23-L40)

The further investigation of what is going on with coverage data is still needed, the issue is not resolved.

**Comment by bdsl** [comment]
@sanmai I've re-run the job and sshd into the docker container at CircleCI. Not sure what details you need though. Maybe these will help:

```
docker@8aa2179df9f7:~$ free -h
              total        used        free      shared  buff/cache   available
Mem:            68G         12G         14G        343M         41G         54G
Swap:            0B          0B          0B

docker@8aa2179df9f7:~$ vmstat
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
10  0      0 8389180 12677688 36648404    0    0   163  2581   25   17 18  6 76  0  0

docker@8aa2179df9f7:~$ swapon -s    
docker@8aa2179df9f7:~$ 

```

**Comment by bdsl** [comment]
Thank you @maks-rafalko . I've subscribed to that PR and will try to give it a go, probably after it's merged to master and maybe not until it's in a release of Psalm - I don't have a lot of time to dedicate to this unfortunately.

Will also try to give running only for changed files a go soon.

Roughly how long did it take you to run Infection for Psalm with the 12GB limit?

**Comment by maks-rafalko** [comment]
> Roughly how long did it take you to run Infection for Psalm with the 12GB limit?

I `ctrl+c`'d it because it was too slow :(

As I said, that PR is just a small starting steps to fix performance issues. Files are being processed too slow because of a big coverage array. We need to run Infection with some profilers and brainstorm about possible ways to improve it. I have some crazy ideas with APCu / Redis, but this is just early thoughts.

**Comment by bdsl** [comment]
@maks-rafalko Thanks again for looking in to this. I finally gave your suggestion of just checking changed files a go, but without much luck, as you can see at https://circleci.com/gh/bdsl/psalm/596 . After that I ran into some bash syntax errors, so for now I'm afraid I've made a PR to stop running infection in the Psalm pipeline. https://github.com/vimeo/psalm/pull/1850

**Comment by sanmai** [comment]
Let's not hurry, please give us a chance. We'll get to the bottom of it. For once we can even start using Circle CI with Infection.

**Comment by bdsl** [comment]
Hi @sanmai . I wasn't trying to hurry you, but I also didn't want to keep an ineffective step in the Psalm pipeline for too long, since it adds complexity for other Psalm contributors wanting to understand that pipeline.

I'm still subscribed to this issue and very happy to give it another go in future.

**Comment by sanmai** [comment]
@bdsl in the master, and in the next version, huge coverage files are no longer an issue.

You may try to integrate Infection with a smaller subset of mutators:

```
{
    ....
    "mutators": {
        "@removal": true,
        "@return_value": true,
        "@zero_iteration": true
    }
}
```
Although I can tell you right away it isn't going to be quick for reasons beyond our control. It should be smooth yet, but not quick.

**Comment by bdsl** [comment]
@sanmai Great, thanks for pinging me. I'm trying it out now with Psalm: https://github.com/bdsl/psalm/commits/start-mutation-testing-again

Assuming this works I'll most likely make a PR to add infection back into the Psalm test pipeline once  thse memory optimisations are in a full release.

**Comment by maks-rafalko** [comment]
@sanmai, shouldn't @bdsl use `--no-progress` & `--only-covered` to get the maximum of your optimizations?

**Comment by bdsl** [comment]
I'm already using ``--only-covered``. Happy to try adding ``--no-progress`` and see how that affects things.

https://github.com/bdsl/psalm/blob/699cb3e284a4abdba3e16686f7fa1d05dc12489e/.circleci/config.yml#L93

**Comment by maks-rafalko** [comment]
more context for `--no-progress` https://github.com/infection/infection/pull/1082

**Comment by bdsl** [comment]
Looks like maybe infection no longer works via a composer global install? https://app.circleci.com/pipelines/github/bdsl/psalm/218/workflows/5dfa1e19-20de-492c-9a58-95c9a03f61f9/jobs/934

**Comment by maks-rafalko** [comment]
I see `Log In to CircleCI` opening this link. Could you please past the result here or open a new issue?

**Comment by bdsl** [comment]
@maks-rafalko Edited my comment above.

**Comment by maks-rafalko** [comment]
ok, extracted to https://github.com/infection/infection/issues/1148

we also have another issue related to `ocramius/package-versions` that might be related https://github.com/infection/infection/issues/876

**Comment by sanmai** [comment]
@bdsl please definitely use `--no-progress --only-covered`

**Comment by sanmai** [comment]
I'm pretty sure there will be no memory-related issues, but lets have @bdsl confirm this is indeed the case.

**Comment by bdsl** [comment]
Psalm isn't well covered with actual unit tests - most features are tested simply by passing samples of PHP code that should have and not have issues detectable by psalm and asserting on the returned set of issues.

We are using pcov.

I have infection part way through a run in an ssh session to circleci right now - so far it's tested just over 400 mutants, so it's looking good.

**Comment by bdsl** [comment]
... and now it's died I'm afraid:

```text
docker@6ac31042464b:~/project$ php -d memory_limit=4G vendor/bin/infection --coverage=build/phpunit --no-progress --only-covered --threads=2

    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Infection - PHP Mutation Testing Framework dev-master@42b722334a3818a9156ff361dc081d23cf5e6528

Running initial test suite...

PHPUnit version: 8.5.2

.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out

....MM.M...M.T....M..M.....MMMMMM..TM...MMM.......   (   50)
..................................................   (  100)
.................T................TT..TT..T.......   (  150)
...........................T......................   (  200)
....................T..T..MM..MTT..M.....MT.......   (  250)
T.......TTTT...T....T.T...T....M.T.TMT.T.T..TTT.T.   (  300)
.........TMMM.TTTT..................T.MM.......TM.   (  350)
..M.M.M..M..TT.....MM..M.M......TT..TMM..........T   (  400)
..M..T...T.T..T
Fatal error: Allowed memory size of 4294967296 bytes exhausted (tried to allocate 4294967304 bytes) in /home/docker/project/vendor/infection/infection/src/Mutator/NodeMutationGenerator.php on line 128

```

**Comment by sanmai** [comment]
Fun times. Just to confirm, are you using a master version for Psalm?

**Comment by bdsl** [comment]
@sanmai Yes, the direct comparison between my branch and Psalm master is https://github.com/vimeo/psalm/compare/master..bdsl:start-mutation-testing-again?expand=1 . Only the circleci config file is changed.

**Comment by bdsl** [comment]
@sanmai Yes, the direct comparison between my branch and Psalm master is https://bit.ly/2Q4LpEx . Only the circleci config file is changed. I had to re-post the comment because the bot is editing the link repeatedly and getting it wrong if I don't use bit.ly

**Comment by bdsl** [comment]
Circle CI only provides 4GB ram in containers on the free plan, so I'm not sure whether this would work if we upped the memory limit. I should be able to try it on my own machine but I need to sleep and do other stuff. Have a good day / night Maks and Alexey, let's talk again soon.

**Comment by sanmai** [comment]
Have a good night, and good morning. It's early morning for me, so no worries here.

Thank you for linking vimeo/psalm#1788.

Now, if you take a closer look at the report:
```
find build/*/ -type f  | xargs ls -sGSrh | tail
```
You'll notice there are several big coverage reports:
```
  69M build/logs/coverage-xml/Psalm/Internal/Type/ParseTree.php.xml
  80M build/logs/coverage-xml/Psalm/Internal/Codebase/Reflection.php.xml
  85M build/logs/coverage-xml/Psalm/Internal/Type/TypeCombination.php.xml
  89M build/logs/coverage-xml/Psalm/Internal/Analyzer/TypeAnalyzer.php.xml
 146M build/logs/coverage-xml/Psalm/Type.php.xml
 163M build/logs/coverage-xml/Psalm/Internal/Analyzer/CommentAnalyzer.php.xml
 171M build/logs/coverage-xml/Psalm/Internal/Codebase/Populator.php.xml
 382M build/logs/coverage-xml/Psalm/Internal/Visitor/ReflectorVisitor.php.xml
```
As a rule of thumb, Infection needs twice the RAM for a given coverage report. E.g. it'll need 800M for 400M report. We see here that a single report shouldn't be a concern with 4GB available.

Before Infection would try to load all reports at once. What Infection does now it loads these reports one by one, while trying to work with them, and it can so happen that Infection has several these reports loaded at once, blowing the memory limit.

Simplest thing here to do is to remove largest of the reports, and see how it'll do:
```
rm build/*/coverage-xml/Psalm/Type.php.xml
rm build/*/coverage-xml/Psalm/Internal/Analyzer/CommentAnalyzer.php.xml
rm build/*/coverage-xml/Psalm/Internal/Codebase/Populator.php.xml
rm build/*/coverage-xml/Psalm/Internal/Visitor/ReflectorVisitor.php.xml
```
This will make corresponding files effectively uncovered, but would let Infection shuffle along.

What we can do here Infection-wise:
- The proper thing to do is to use a different, streaming, XML parser. There are several of them. It's a big job, and I wish I was mistaken here.
- As a workaround, we can control for memory limit, and do not load queue further jobs.
   - `ParallelProcessRunner` controls the queue, but if done inside `ParallelProcessRunner` this'll make things slower.
   - `IndexXmlCoverageParser` can postpone loading certain files if it knows there's no way to fit them into memory. But this will require it to either pass a No-Op mutations, or communicate with `ParallelProcessRunner` thought events to trigger `freeTerminatedProcesses`.

**Comment by sanmai** [comment]
With Psalm there's another problem: there are hidden dependencies between tests. Therefore Infection would not reliably detect all mutations. This is a much bigger problem, so I guess we'll have to just keep that in mind. Verify with:

    php vendor/bin/phpunit --order-by=random --stop-on-defect

If it fails, means there's an undeclared dependency.

**Comment by theofidry** [comment]
@sanmai can it be that we have a leak still?

**Comment by sanmai** [comment]
@theofidry I was able to run Infection [with a smaller subset of profiles](https://github.com/bdsl/psalm/pull/2#issuecomment-597426322).

**Comment by maks-rafalko** [comment]
we did many changes to run Infection for big projects like Psalm, I don't think we need to keep it open.

Really big projects can/should run Infection for a diff (modified and/or added files) only
